File: 031.NextPermutation/MarkCSIE/NextPermutation/src/NextPermutation.java

import java.util.Arrays;

public class NextPermutation {

    public static void main(String[] args) {
        int[] testCase1 = new int[]{1, 2, 3};
        System.out.print(Arrays.toString(testCase1) + "->");
        NextPermutation test = new NextPermutation();
        test.nextPermutation(testCase1);
        System.out.println(Arrays.toString(testCase1));
    }

    public void nextPermutation(int[] nums) {
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                for (int j = nums.length - 1; j >= i + 1; j--) {
                    if (nums[j] > nums[i]) {
                        int temp = nums[i];
                        nums[i] = nums[j];
                        nums[j] = temp;

                        // reverse
                        int start = i + 1;
                        int end = nums.length - 1;
                        while (start < end) {
                            temp = nums[start];
                            nums[start] = nums[end];
                            nums[end] = temp;
                            start++;
                            end--;
                        }
                        return;
                    }
                }
            }

        }

        // reverse
        int start = 0;
        int end = nums.length - 1;
        while (start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
}

File: 031.NextPermutation/wyrj/NextPermutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var swap = function(arr, i, j) {
    var temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}
var reverse = function(arr, start, end) {
    start = start || 0;
    end = end || arr.length - 1;
    while(start < end) {
        swap(arr, start, end);
        start += 1;
        end -= 1;
    }
}
var nextPermutation = function(nums) {
    var len = nums.length;
    var i, j;
    for (i = len - 1; i > 0; i--) {
        if (nums[i] > nums[i - 1]) {
            for (j = len - 1; j >= i; j--) {
                if (nums[j] > nums[i - 1]) {
                    swap(nums, i - 1, j);
                    reverse(nums, i);
                    return;
                }
            }
        }
    }
    reverse(nums);
};